### ISSUE TITLE
Merge nested

### ISSUE THREAD

**Issue by MarioVanDenEijnde** [agershun/alasql#551]
Hello alasql team,

How can we merge json objects that include nested objects?
We tried JOIN and EXTEND and UNION but cannot seem to get the right result

Please advise.

```
res1: [{"inspecteur":"Jan","keuring":[{"keuring":"Keuring 03","inspecteur":"Jan","date":"2016-01-22"},{"keuring":"Keuring 04","inspecteur":"Jan","date":"2016-01-20"},{"keuring":"Keuring 01","inspecteur":"Jan","date":"2016-01-18"}]},{"inspecteur":"Piet","keuring":[{"keuring":"Keuring 02","inspecteur":"Piet","date":"2016-01-20"}]}]
```
```
res2: [{"inspecteur":"Jan","keuring":[{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"}]},{"inspecteur":"Piet","keuring":[{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"}]}]
```
result should be:
```
res3: [{"inspecteur":"Jan","keuring":[{"keuring":"Keuring 01","inspecteur":"Jan","date":"2016-01-18"},{"date":"2016-01-19"},{"keuring":"Keuring 04","inspecteur":"Jan","date":"2016-01-20"},{"date":"2016-01-21"},{"keuring":"Keuring 03","inspecteur":"Jan","date":"2016-01-22"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"}]},{"inspecteur":"Piet","keuring":[{"date":"2016-01-18"},{"date":"2016-01-19"},{"keuring":"Keuring 02","inspecteur":"Piet","date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"}]}]
```

**Comment by agershun** [comment]
You try this statement:
```sql
    var res3 = alasql('SELECT a.inspecteur, a.keuring->concat(b.keuring) AS keuring \
            FROM ? AS a OUTER JOIN ? as b USING inspecteur',[res1,res2]);
```
Here we JOIN two tables by ```inspecteur``` table.

To merge two a.keuring fields I use JavaScript ```concat()``` function.

The result is:
```json
[{"inspecteur":"Jan","keuring":[{"keuring":"Keuring 03","inspecteur":"Jan","date":"2016-01-22"},{"keuring":"Keuring 04","inspecteur":"Jan","date":"
2016-01-20"},{"keuring":"Keuring 01","inspecteur":"Jan","date":"2016-01-18"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"da
te":"2016-01-21"},{"date":"2016-01-22"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22
"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01-21"},{"date":"2016-01-22"}]},{"inspecteur":"Piet","keuring":[
{"keuring":"Keuring 02","inspecteur":"Piet","date":"2016-01-20"},{"date":"2016-01-18"},{"date":"2016-01-19"},{"date":"2016-01-20"},{"date":"2016-01
-21"},{"date":"2016-01-22"}]}]
```

Could you test it?

**Comment by MarioVanDenEijnde** [comment]
Hello Andrey,

The result is not what we need. Basically in res2 we have an array of date objects and we need to replace these objects with the objects in res1 (if exist)
Your suggestion only adds the objects.
Maybe we could use something like: IIF(b.keuring == null,a.keuring,b.keuring) AS keuring
(my suggestion is not yet working though)

Kind regards,
Mario

**Comment by agershun** [comment]
May be this:
``` 
    IIF(LEN(b.keuring)>0,a.keuring,b.keuring) AS keuring
```

**Comment by MarioVanDenEijnde** [comment]
No. But I think we miss out that we want to extend or replace the objects inside the array "keuring" and not the array itself.

**Comment by agershun** [comment]
So you want to concatinate these arrays and then remove duplicated objects, is this correct?

**Comment by MarioVanDenEijnde** [comment]
Hello Andrey,

Basically in res2 we have an array of date objects inside an array.
And we need to replace these objects with the objects in res1 (if exist)
If you would take a look at the requested result in the first post I think you get what we need.

Kind regards,
Mario

**Comment by MarioVanDenEijnde** [comment]
I thought mayb I can get away just adding the date objects into the array like below and later disable rendering the empty ones:
But the code below code does not do this. Can you advise?
```
var res3 = alasql('SELECT a.inspecteur, ARRAY(a.keuring._)->concat(ARRAY(b.keuring._)) AS keuring FROM ? AS a OUTER JOIN ? as b USING inspecteur',[res1,res2]);
```

**Comment by agershun** [comment]
I think we can do it in two steps:
1) join two arrays with two fields ```akeuring``` and ```bkeuring```
```sql
var res = alasql('SELECT a.inspecteur, a.keuring AS akeuring, b.keuring AS bkeuring FROM ? AS a OUTER JOIN b USING inspecteur');
```
Then walk on ```res``` array and join subarrays like:
```
res.forEach(function(d){
    d.keuring = alasql'SELECT [date],[inspecteur],keuring FROM ? AS a JOIN ? AS b USING [date]', [d.akeuring, d.bkeuring]);
    delete d.akeuring;
    delete d.bkeuring;
});
```

This is only idea, I did not tested it.

**Comment by agershun** [comment]
Here is the code:
```
var res = alasql('SELECT a.inspecteur, a.keuring AS akeuring, b.keuring AS bkeuring FROM ? AS a OUTER JOIN ? b USING inspecteur',[res1,res2]);

res.forEach(function(d){
    d.keuring = alasql('SELECT (a.[date] OR b.[date]) AS [date],\
         (a.[inspecteur] OR b.[inspecteur]) AS inspecteur,\
         (a.keuring OR b.keuring) AS keuring \
         FROM ? AS a OUTER JOIN ? AS b USING [date]', [d.akeuring||[], d.bkeuring||[]]);
    delete d.akeuring;
    delete d.bkeuring;
});
```
and here - results:
```json
[{"inspecteur":"Jan","keuring":[{"date":"2016-01-22","inspecteur":"Jan","keuring":"Keuring 03"},{"date":"2016-01-22","inspecteur":"Jan","keuring":"
Keuring 03"},{"date":"2016-01-22","inspecteur":"Jan","keuring":"Keuring 03"},{"date":"2016-01-20","inspecteur":"Jan","keuring":"Keuring 04"},{"date
":"2016-01-20","inspecteur":"Jan","keuring":"Keuring 04"},{"date":"2016-01-20","inspecteur":"Jan","keuring":"Keuring 04"},{"date":"2016-01-18","ins
pecteur":"Jan","keuring":"Keuring 01"},{"date":"2016-01-18","inspecteur":"Jan","keuring":"Keuring 01"},{"date":"2016-01-18","inspecteur":"Jan","keu
ring":"Keuring 01"},{"date":"2016-01-19"},{"date":"2016-01-21"},{"date":"2016-01-19"},{"date":"2016-01-21"},{"date":"2016-01-19"},{"date":"2016-01-
21"}]},{"inspecteur":"Piet","keuring":[{"date":"2016-01-20","inspecteur":"Piet","keuring":"Keuring 02"},{"date":"2016-01-18"},{"date":"2016-01-19"}
,{"date":"2016-01-21"},{"date":"2016-01-22"}]}]
```

**Comment by agershun** [comment]
Probably, we need to change something in the second ```JOIN``` to remove this record: ```{"date":"2016-01-22","inspecteur":"Jan","keuring":"Keuring 03"}```. May be INNER join?

**Comment by mathiasrw** [comment]
@MarioVanDenEijnde input for future issues: Its best if you upfront state what you are hoping to get as result - that way we avoid a lot of back and forth.

**Comment by MarioVanDenEijnde** [comment]
Hello Andrey,
The doubles was my mistake in res2 I removed them. And thanks a lot: this works.

Hello Mathias,
I thought I covered this in the first post: "result should be"

Again thanks a lot.
Kind regards,
Mario

**Comment by mathiasrw** [comment]
@MarioVanDenEijnde Sorry for that. I missed it, and only saw the many "back and forth"s. I withdraw my comment !

**Comment by agershun** [comment]
https://www.youtube.com/watch?v=Zf052uxFF58&feature=youtu.be&t=58s :)

**Comment by mathiasrw** [comment]
Indeed you did !

**Comment by MarioVanDenEijnde** [comment]
Hello Guys,

Are you interested in another piece of the puzzle?
This is what we got:
```
[{"title":"Keuring 03","inspecteur":"Jan","date":"2016-01-22"},
{"title":"Keuring 04","inspecteur":"Jan","date":"2016-01-20"},
{"title":"Keuring 02","inspecteur":"Piet","date":"2016-01-20"},
{"title":"Keuring 05","inspecteur":"Jan","date":"2016-01-20"},
{"title":"Keuring 01","inspecteur":"Jan","date":"2016-01-18"}]
```
This is the result we need:
```
[
	{	"inspecteur":"Jan",
		dates:
			[
				{	"date":"2016-01-22",
					keuringen:
						[
							{	"title":"Keuring 03"
							}
						]
				},
				{	"date":"2016-01-20",
					keuringen:
						[
							{	"title":"Keuring 04"
							},
							{	"title":"Keuring 05"
							}
						]
				},
				{	"date":"2016-01-18",
					keuringen:
						[
							{	"title":"Keuring 01"
							}
						]
				}
			]
	},
	{	"inspecteur":"Piet",
		dates:
			[
				{	"date":"2016-01-20",
					keuringen:
						[
							{	"title":"Keuring 02"
							}
						]
				}
			]
	}
]
```

Kind regards,
Mario

**Comment by MarioVanDenEijnde** [comment]
Ah that is just a double GROUP:
```
var resX = alasql('SELECT inspecteur, date, ARRAY(res._) AS keuring FROM ? res GROUP BY inspecteur,date', [res]);
```
Stupid ME,

Kind regards,
Mario

**Comment by agershun** [comment]
:+1: Please, do not forget to put ```[date]``` into square brackets (```date``` is (heavily overloaded) reserved keyword in AlaSQL).

**Comment by mathiasrw** [comment]
Is it super crazy to suggest that we try to reshape the jison to let all reserved keywords that MUST be followet by `(....)` to react only on `/KEYWORD *(/` and dance around with the rest of the code to handle that there is no starting `(`token for "function" keywords ?

**Comment by agershun** [comment]
@mathiasrw Can you an example to explain the idea?

**Comment by mathiasrw** [comment]
A lot of keywords as `min` `max` are only present if a `(` follows. So to be able to select a row named `min` we need the keyword to do a lookforward if a `(` comes after. 

I have been playing with it and found that [JISON supports lookforward](https://github.com/zaach/jison/wiki/Deviations-From-Flex-Bison#negative-lookahead), so by replacing 

`"MIN" return 'MIN'`

with

`'MIN'(\s+)?/'(' return 'MIN'`

we can now

`SELECT min(min) as min FROM @[{min:5},{min:3}]` and it returns `[{min:3}]`

I have pushed the changes. It should be the same with `MAX` but found that the string `max` is also a keyword you can use when setting size of table field - so at the moment you must do a 

`SELECT max([max]) as max FROM @[{max:5},{max:3}]`=> `[ { max: 5 } ]`

But I still see it as an improvement as you will not get a parse error when doing a "normal" 

`SELECT max FROM @[{max:5},{max:3}]` => [ { max: 5 }, { max: 3 } ].

I will be poking a bit more around what other function keyword we can save the same way. 

Question: I see no token named `DATE` - how come its a keyword?

By the way: JISON supports that $1 is the same as $NAMEOFTHETOKEN so we can write like this:

```
ColumnType
	: LITERAL LPAR NumberMax COMMA NUMBER RPAR
		{ $$ = {dbtypeid: $LITERAL, dbsize: $NumberMax, dbprecision: +$NUMBER} }
	| LITERAL LITERAL LPAR NumberMax COMMA NUMBER RPAR
		{ $$ = {dbtypeid: $LITERAL1+($LITERAL2?' '+$LITERAL2:''), dbsize: $NumberMax, dbprecision: +$NUMBER} }
```

(notice that when more with same name they get anumber in the end (like `LITERAL`))

**Comment by agershun** [comment]
1. ```MIN``` and ```MAX``` are aggregators. They are not keyword, but if the looks like functions (followed by ```( )```), then AlaSQL interpret them as aggregators.

2. The ```DATE``` and ```Date```  is a keyword for datatypes and functions also. It is defined via:
```js
    alasql.fn.Date = Date;
```

3. Yes, we can. I also learned it after I finished most part of the parser. Anyway, it should be reqritten, and we can use this approach to minimize the size of the code and to increase durability.
